NDTensors/src/lib/TypeParameterAccessors/test/test_basics.jl

```diff
@@ -0,0 +1,77 @@
+@eval module $(gensym())
+using Test: @test_throws, @testset
+using NDTensors.TypeParameterAccessors:
+  TypeParameterAccessors,
+  Position,
+  TypeParameter,
+  set_parameter,
+  set_parameters,
+  specify_parameter,
+  specify_parameters,
+  type_parameter,
+  type_parameters,
+  unspecify_parameter,
+  unspecify_parameters
+include("utils/test_inferred.jl")
+@testset "TypeParameterAccessors basics" begin
+  @testset "Get parameters" begin
+    @test_inferred type_parameter(Array{Float64}, 1) == Float64 wrapped = true
+    @test_inferred type_parameter(Array{Float64}, Position(1)) == Float64
+    @test_inferred type_parameter(Val{3}) == 3
+    @test_throws ErrorException type_parameter(Array, 1)
+    @test_inferred type_parameter(Array{Float64}, eltype) == Float64
+    @test_inferred type_parameter(Matrix{Float64}, ndims) == 2
+    @test_throws ErrorException type_parameter(Array{Float64}, ndims) == 2
+    @test_inferred type_parameters(Matrix{Float64}, (2, eltype)) == (2, Float64) wrapped =
+      true
+    @test_inferred type_parameters(Matrix{Float64}, (Position(2), eltype)) == (2, Float64)
+  end
+  @testset "Set parameters" begin
+    @test_inferred set_parameter(Array, 1, Float64) == Array{Float64} wrapped = true
```

Comment:
does this fail if `wrapped = false`

Reply:
Yes, type inference for most of these functions with runtime inputs (like the `1` here) fail without wrapping into a function call. I think that is just how Julia's compiler works:
```julia
julia> @code_warntype (Float32, Float64, 3)[2]
MethodInstance for getindex(::Tuple{DataType, DataType, Int64}, ::Int64)
  from getindex(t::Tuple, i::Int64) @ Base tuple.jl:31
Arguments
  #self#::Core.Const(getindex)
  t::Tuple{DataType, DataType, Int64}
  i::Int64
Body::Union{Int64, DataType}
1 ─      nothing
│   %2 = Base.getfield(t, i, $(Expr(:boundscheck)))::Union{Int64, DataType}
└──      return %2


julia> @code_warntype (() -> (Float32, Float64, 3)[2])()
MethodInstance for (::var"#6#7")()
  from (::var"#6#7")() @ Main REPL[16]:1
Arguments
  #self#::Core.Const(var"#6#7"())
Body::Type{Float64}
1 ─ %1 = Core.tuple(Main.Float32, Main.Float64, 3)::Core.Const((Float32, Float64, 3))
│   %2 = Base.getindex(%1, 2)::Core.Const(Float64)
└──      return %2
```
I believe that technically the `1` is a global variable, so then it can't be turned into a `Const` by the compiler.